testgen/src/light_block.rs

```diff
@@ -0,0 +1,252 @@
+use std::str::FromStr;
+use tendermint::block::signed_header::SignedHeader;
+use tendermint::node::Id as PeerId;
+use tendermint::validator;
+use tendermint::validator::Set as ValidatorSet;
+
+/// A light block is the core data structure used by the light client.
+/// It records everything the light client needs to know about a block.
+/// NOTE: This struct & associated `impl` below are a copy of light-client's `LightBlock`.
+/// The copy is necessary here to avoid a circular dependency.
+/// Cf. https://github.com/informalsystems/tendermint-rs/issues/605
+/// TODO: fix redundant code without introducing cyclic dependency.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct TMLightBlock {
+    /// Header and commit of this block
+    pub signed_header: SignedHeader,
+    /// Validator set at the block height
+    pub validators: ValidatorSet,
+    /// Validator set at the next block height
+    pub next_validators: ValidatorSet,
+    /// The peer ID of the node that provided this block
+    pub provider: PeerId,
+}
+
+impl TMLightBlock {
+    /// Constructs a new light block
+    pub fn new(
+        signed_header: SignedHeader,
+        validators: ValidatorSet,
+        next_validators: ValidatorSet,
+        provider: PeerId,
+    ) -> Self {
+        Self {
+            signed_header,
+            validators,
+            next_validators,
+            provider,
+        }
+    }
+}
+
+/// We use this data structure as a simplistic representation of LightClient's LightBlock
+#[derive(Debug, Options, Deserialize, Clone)]
+pub struct LightBlock {
+    #[options(help = "header (required)", parse(try_from_str = "parse_as::<Header>"))]
+    pub header: Option<Header>,
+    #[options(help = "commit (required)", parse(try_from_str = "parse_as::<Commit>"))]
+    pub commit: Option<Commit>,
+    #[options(
+        help = "validators (required), encoded as array of 'validator' parameters",
+        parse(try_from_str = "parse_as::<Vec<Validator>>")
+    )]
+    pub validators: Option<Vec<Validator>>,
+    #[options(
+        help = "next validators (default: same as validators), encoded as array of 'validator' parameters",
+        parse(try_from_str = "parse_as::<Vec<Validator>>")
+    )]
+    pub next_validators: Option<Vec<Validator>>,
+    #[options(help = "peer id (default: peer_id())")]
+    pub provider: Option<String>,
+}
+
+impl LightBlock {
+    /// Constructs a new Testgen-specific light block
+    pub fn new(
+        header: Header,
+        commit: Commit,
+        validators: Vec<Validator>,
+        next_validators: Vec<Validator>,
+        provider: &str,
+    ) -> Self {
+        Self {
+            header: Some(header),
+            commit: Some(commit),
+            validators: Some(validators),
+            next_validators: Some(next_validators),
+            provider: Some(provider.to_string()),
+        }
+    }
+
+    pub fn new_default(validators: &[Validator], height: u64) -> Self {
```

Comment:
Is it possible to have a function with the same signature as this one, except it's missing the input parameter `validators: &[Validator]`? I mean something of this type:

```rust
    pub fn new_default(height: u64) -> Self
```

Reply:
done ✅